### DISCUSSION TITLE
Tox 3 s905x4 4/32 arm64

### DISCUSSION THREAD

**Discussion by x3emal** [devmfc/debian-on-amlogic#25] [Q&A] [answered]
Hello add plz boot.config for TOX 3 amlogic s905x4

**Comment by devmfc** [comment, marked as answer]
Sure, my pleasure, just send me a TOX3 box or the money to buy one and I'll make one for you.

**Reply by mathew500** [reply]
Hi,
Is the hardware a requirement or is a image dump of CoreElec sufficient?
Or i can provide the dtb.img from the CoreElec project that works. Or a .dtb/.dts dump from a working image.

Best regards Mathias

**Reply by devmfc** [reply]
If it boots with one of the generic configs, probably not. 
Or when someone makes a mainline DTB version for the box,  probably not
But otherwise I need UART (serial) access to the box, so I need physical acces to the box.

CoreElec is based on the old (and buggy) vendor kernel and is not compatible with mainline linux, so no, a Coreelec image does not help.

**Reply by mathew500** [reply]
Hi, 

I know this is a couple of months old, however, how much money would you need to add the box? Don't know where you're living and what the costs would be for you, hence the question.

Best regards Mathias

**Reply by devmfc** [reply]
I’m in Europe and these boxes are way too expensive for what you get (around 70 euro’s on AliExpress). Yes, the Android software is a little better (based on my experience with a Tox1) than the standard boxes, but still too expensive.

But even then.. I own a Tox1, it uses secure boot, so the bootloader is signed and encrypted. Probably the Tox3 has that as well. That makes it harder to boot another image. I got it to work with Tox1, but the risk of bricking such a box is much higher.

I will probably get the box to boot, but it has a S905X4 (SC2) SoC, so there is NO hdmi support, no 100M ethernet support at his moment with a mainline kernel. And I don’t have the knowledge to fix that in an acceptable amount of time. That means: only useable as a minimal Linux server, no graphical desktop.

So it will be much cheaper for you to buy a supported cheap S905X3 box with gigabit ethernet and just use that. Depending of what you want with it of course.

Best regards

**Reply by mathew500** [reply]
Hi,

From Europe as well.
Thank you for your honest response.
Booting CoreELEC is easy, so not to sure about the booting difficulties, but I hear what you're saying.
Was just hoping to make some use of it, now that I have bought it.
As you recommend, it's probably smarter to go the S905X3 route.

**Comment by AbraxasDMG** [comment]
Hi everyone. I've been trying to get this box into an acceptable working condition (at least according to my needs).

I have used the Vontar x4 gigabit dtb that devmfc's build already contained, since nothing significant was failing with it (I run the box as the headless server). Main pain point was the lack of network connectivity. The front led panel (vfd) was also displaying rubbish.
Below are the details relevant to how I fixed both of these points.

Original Vontar x4 dtb is modified in the segment relevant to the ethernet phy. I have used the values from the box's original android-provided dtb to populate this.

Aside from dtb, an out-of-tree driver is needed (the gbe phy on the box is Maxio MAE0621A). I have used the one I've found in one of the CoreELEC builds (CoreELEC-Amlogic-ne.aarch64-21.2-Omega.tar).
That driver had to be modified (mostly a guess-work. didn't know what I was doing.), in order to be able to bind it to the device, since Maxio phy is not identifying itself correctly in the system (phy_id reported is 0xffff4411 instead of 0x7b744411).

In order to avoid having to explicitly run build for each new kernel version, I tried to set this up as a dkms-managed driver.

Another problem that appeared then is that the driver module would not be loaded early enough, which would still result in a phy failure. That could be resolved by manually loading the module and manually bouncing the interface. In order to avoid having to do that, I've added few lines to the module-loading-related configuration files, which seem to do the job.

Lastly, the vfd mapping is not adequate for tox3 front led panel. Even the "load" message at the start is not displayed properly.
I've found out that the display works correctly when the map1 from /usr/lib/tm16xx-display/tm16xx-map.sh is applied to it, so I have modified the /usr/lib/tm16xx-display/display_service.sh to activate that mapping. The "load" message at the boot is still not shown properly, but I can live with that. As soon as tm16xx service starts, the display shows meaningful info.

<u>This is what worked for me. It might work for you too, but you might encounter issues. I can't provide any guarantee or assume responsibility.</u>



## <u>Boot configuration and dtb</u>

Add the following line to the /boot/boot.config file:
`box=tox3_gigabit`
Make sure that all other lines that start with "box=" are commented out.

Copy box-tox3_gigabit.config to the /boot/box-config directory.

Copy meson-sc2-tox3-gbit.dtb to the /boot/dtb-X.YY.ZZ-meson64/amlogic directory.



## <u>Maxio driver</u>

Required packages on the box:
```
dkms
build-essential
linux-headers-$(uname -r)
```

Create /usr/src/maxio-1.0 directory and copy maxio.c, dkms.conf and Makefile files to it.
Change to /usr/src/maxio-1.0directory and then execute:
```
sudo dkms add -m maxio -v 1.0
sudo dkms build -m maxio -v 1.0 -k "(uname -r)"
sudo dkms install -m maxio -v 1.0 -k "(uname -r)"
```

Reload the driver.
```
sudo modprobe -r maxio 2>/dev/null || true
sudo modprobe maxio
```

Bounce the interface so the MDIO probe runs again.
```
sudo ip link set eth0 down
sudo ip link set eth0 up
```

Check the interface, connection and the binding.
```
ethtool eth0
ip a
dmesg | egrep -i 'maxio|phy|stmmac|link'
```

Create /etc/modprobe.d/99-maxio.conf and add the following lines to it:
```
softdep meson8b-dwmac pre: maxio
softdep stmmac_platform pre: maxio
```

Add `maxio` ina new line to the /etc/modules-load.d/modules.conf.



That should be all. I hope I didn't forget anything.
Good luck!
[tox3.zip](https://github.com/user-attachments/files/22662989/tox3.zip)

**Reply by devmfc** [reply]
Nice work and thanks for reporting! 
If I have some time before the next release I will try to add your findings to the default image.

**Comment by u0543** [comment]
Dear AbraxasDMG , how did you install linux-headers-$(uname -r)?

**Reply by AbraxasDMG** [reply]
Hi. Just try executing `sudo apt install linux-headers-$(uname -r)`.

**Reply by elopukki** [reply]
> Уважаемый AbraxasDMG, как вы установили linux-headers-$(uname -r)?

Using the example of my setup. Substitute your own 
Download linux-headers-12/6/30-meson64_20250522_arm64.deb  
Choose your own versions.
for example, copy to the home folder and go to it (you can delete the file after installing the drivers)
don't forget to first apt update && apt upgrade -y
cd /home  
run the command 
dpkg -i linux-headers-6.12.30-meson64_20250522_arm64.deb 
an error will pop up. You need to run
apt --fix-broken install
and repeat the above command again. ( Maybe it's the other way around, but I've been going for 4 days and I don't want to double-check since I'm on linux --0 )
And you can install drivers.
The only thing is that I then directly specified the headers during the build. 
dkms build -m maxio -v 1.0 --kernelsourcedir=/usr/src/linux-headers-6.12.30-meson64
dkms install -m maxio -v 1.0 --kernelsourcedir=/usr/src/linux-headers-6.12.30-meson64

<img width="915" height="663" alt="Снимок экрана 2026-04-02 203125" src="https://github.com/user-attachments/assets/56290098-a3e7-42af-9c45-eed812e72fc9" />





But I didn't fix the display, I couldn't. Shows the cosmos

**Reply by AbraxasDMG** [reply]
> Dear AbraxasDMG , how did you install linux-headers-$(uname -r)?

Sorry, this was a while ago. Now I got it what was asked.
Just install the headers that are identical version to your running kernel version.

You can figure out which would be the right package by executing the following in your terminal:
`echo linux-headers-$(uname -r)`

Then you should browse the Release section here and find the deb file, which you can install with either `sudo dpkg -i linux-headers-<version>-meson64_<date>_arm64.deb` or `sudo apt install ./linux-headers-<version>-meson64_<date>_arm64.deb`.

**Comment by u0543** [comment]
Hi AbraxasDMG, thank you so much for your prompt and detailed response.
I've managed to apply your patch.
The only thing I am missing now is booting from USB (I'm able to boot from SD-cardonly ) or mounting a SSD disk partition as a root FS (fstab update doesn't do the trick).

**Reply by AbraxasDMG** [reply]
Hi. Glad to help.

I remember there was an issue with setting up multiboot, but I can't remember what have I done exactly.
There was an issue with aml-multiboot-setup.sh script identifying the partitions or could not deal with their current scheme, something like that. I ended up following the script line by line in order to figure out where is it failing and what to do to achieve expected behavior without changing the script itself (to avoid having to modify it on every release).

I don't know if you're facing that same issue or some other/additional one. I'd suggest you create another post for that specific issue. You might get others looking at it and advising much quicker than here in Tox 3 device specific thread (which people might skip due to not having any touch with it).

Good luck!

**Reply by elopukki** [reply]
./aml-multiboot-setup.sh
then
./aml-install-to-emmc.sh
transferred to the internal memory

Lastly, the vfd mapping is not adequate for tox3 front led panel. Even the "load" message at the start is not displayed properly.
I've found out that the display works correctly when the map1 from /usr/lib/tm16xx-display/tm16xx-map.sh is applied to it, so I have modified the /usr/lib/tm16xx-display/display_service.sh to activate that mapping
tell me what was changed in /usr/lib/tm16xx-display/display_service.sh what would apply map1 ?
or you need to rebuild meson-sc2-tox3-gbit.dts for the right core

**Reply by AbraxasDMG** [reply]
> tell me what was changed in /usr/lib/tm16xx-display/display_service.sh what would apply map1 ? or you need to rebuild meson-sc2-tox3-gbit.dts for the right core

I did a quick and dirty fix for my own use. Just edited the display_service.sh by overwriting the $dev_path/mapping with the map1 hardcoded map1 code.

<img width="631" height="292" alt="image" src="https://github.com/user-attachments/assets/13f3d4a5-ce70-40d5-835c-8fb21060207c" />

**Reply by elopukki** [reply]
It worked! Thanks!

**Comment by u0543** [comment]
Hi AbraxasDMG, I've just realized I booted Debian with  /boot/boot.config file set to vontarx4!
Your LAN patch was applied and works under this environment.
When I noticed I missed to copy TOX3 dtb, copied it into dtb dir and copied box-tox3_gigabit.config to the /boot/box-config directory I was unable to boot the box with /boot/boot.config file set to tox3 ;-(
BTW, LCD displays not the FULL rubbish. It gives the last 3 difgits of the IP address in the reverse order (132 stands for 231 in my case).

**Reply by AbraxasDMG** [reply]
Hi. I can only speak for the combination of the image and kernel I've built this on: Ubuntu 24.04 (Noble) + kernel v6.12.
I can't be certain if different Ubuntu/Debian images could render some of this solution invalid. Sorry.

If you've placed/set everything in the right place (as described [here](https://github.com/devmfc/debian-on-amlogic/discussions/25#discussioncomment-14575524), and as packed within the attached tox3.zip archive at the end of it), it should work without needing any further modification.

Try to validate the paths of different files relevant to this topic, and the values in different configuration files (make sure that they are the same, not only similar). That's the best I can do, sorry.

**Comment by u0543** [comment]
Hi AbraxasDMG,
Thank you.

Looks like your patch is valid for bookworm versions.
Under trixie I get:
dkms build -m maxio -v 1.0 --kernelsourcedir=/usr/src/linux-headers-6.18.10-meson64
The kernel is built without module signing facility, modules won't be signed

Building module(s).....(bad exit status: 2)
Failed command:
make -j4 KERNELRELEASE=6.18.10-meson64 -C /usr/src/linux-headers-6.18.10-meson64 M=/var/lib/dkms/maxio/1.0/build modules

Error! Bad return status for module build on kernel: 6.18.10-meson64 (aarch64)
Consult /var/lib/dkms/maxio/1.0/build/make.log for more information
I did not get this error under bookworm.

**Reply by elopukki** [reply]
> Hi AbraxasDMG, Thank you.
> 
> Looks like your patch is valid for bookworm versions. Under trixie I get: dkms build -m maxio -v 1.0 --kernelsourcedir=/usr/src/linux-headers-6.18.10-meson64 The kernel is built without module signing facility, modules won't be signed
> 
> Building module(s).....(bad exit status: 2) Failed command: make -j4 KERNELRELEASE=6.18.10-meson64 -C /usr/src/linux-headers-6.18.10-meson64 M=/var/lib/dkms/maxio/1.0/build modules
> 
> Error! Bad return status for module build on kernel: 6.18.10-meson64 (aarch64) Consult /var/lib/dkms/maxio/1.0/build/make.log for more information I did not get this error under bookworm.

I installed the drivers on 6.18.10-meson64 and they work, but the kernel did not suit me for homassistant.

**Comment by u0543** [comment]
BTW, Debian does boot from USB.
Ir didn't want to boot from one of card readers plugged into USB, but it boots from a real usb pendrive.

**Reply by vtvr5box-cell** [reply]
Tox3 2/16. I soldered out the emmc because it went to RO. Is there an image I can write to SD to load and use Coreelec? The UART log shows that Coreelec cannot be loaded from SD
